text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+  fn bar(x: u32) {
+    let a = |x| { x + 1 }; // local name: 0C
+    let b = |x| { x + 2 }; // local name: 0Cs_
+
+    a(b(x))
+  }
+}
+
+```
+
+In the above example we have two closures, the one assigned to `a` and the one assigned to `b`. The first one would get the local name `0C` and the second one the name `0Cs_`. The `0` signifies the length of their (empty) name. The `C` is the namespace tag, analogous to the `V` tag for the value namespace. The `s_` for the second closure is the disambiguation index (index `0` is, again, encoded by not appending a suffix). Their full names would then be `N15mycrate_4a3b56d3foo3barV0CE` and `N15mycrate_4a3b56d3foo3barV0Cs_E` respectively.
+
+
+### Methods
+
+Methods are nested within `impl` or `trait` items. As such it would be possible to construct their symbol names as paths like `my_crate::foo::{{impl}}::some_method` where `{{impl}}` somehow identifies the the `impl` in question. Since `impl`s don't have names, we'd have to use an indexing scheme like the one used for closures (and indeed, this is what the compiler does internally). Adding in generic arguments to, this would lead to symbol names looking like `my_crate::foo::impl'17::<u32, char>::some_method`.
+
+However, in the opinion of the author these symbols are very hard to map back to the method they represent. Consider a module containing dozens of types, each with multiple `impl` blocks generated via `#[derive(...)]`. In order to find out which method a symbol maps to, one would have to count the number of handwritten and macro generated `impl` blocks in the module, and hope that one correctly guessed the number of `impl` blocks introduced by the given derive-macro (each macro invocation can introduce `0..n` such blocks). The name of the method might give a hint, but there are still likely to be dozens of methods named `clone`, `hash`, `eq`, et cetera.
+
+The RFC therefore proposes to keep symbol names close to how methods are represented in error messages, that is:
+
+- `Foo<u32, char>::some_method` for inherent methods, and
+- `<Foo<u32, char> as SomeTrait<Quux>>::some_method` for trait methods.
+
+This can be achieved by extending the definition of absolute paths that we have used so far. Instead of the path root always being a crate-id, we now also allow a path to start with a single type (i.e. the self-type of an inherent method) or with a pair of self-type and the trait being implemented. The kind of root is indicated by the first character of the `N` starting the path:
+
+- a decimal digit signifies a path with a crate-id root (since crate-ids always start with a digit),
+- an `M` signifies a path with a single type as its root, and
+- an `X` signifies a path with a self-type/trait pair as its root.
+
+Thus, this extended form of paths would have the following syntax:
+
+```
+<abs-path>  := N <path-root> <ident>* [I <generic-arguments> E] E
+
+<path-root> := <crate-id>
+            |  M <type>
+            |  X <type> <abs-path>
```

Comment:
Bringing this comment up again: https://internals.rust-lang.org/t/pre-rfc-a-new-symbol-mangling-scheme/8501/4?u=rpjohnst. Would it make sense to move the trait's self type into its argument list? It reorders things from how they are displayed in error messages, but simplifies the grammar a bit.

Reply:
I agree, we already treat `<X as Trait<Y, Z>>` as sugar for `Trait` applied with `[X, Y, Z]`, there's no real reason to have it separate here.

Reply:
Oh, wait, an on-the-fly demangler needs to have everything in demangled order, right.
Is this only needed for `<X as Trait<Y, Z>>`, or are there other "out of order" constructs?

Reply:
As far as I can tell `<X as Trait<Y, Z>>` is the only case.